File: lintcode/swap_nodes_in_pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        current = dummy = ListNode(None)
        dummy.next = head
        while current.next and current.next.next:
            first, second = current.next, current.next.next
            current.next, first.next, second.next = second, second.next, first
            current = current.next.next
        return dummy.next

File: leetcode/swap_nodes_in_pairs.js
/* Given a linked list, swap every two adjacent nodes and return its head.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    current = dummy = new ListNode(0)
    dummy.next = head
    for (; current.next && current.next.next; current = current.next.next) {
        first = current.next
        second = current.next.next
        current.next = second
        first.next = second.next
        second.next = first
    }
    return dummy.next
};
